build/ci.go

```diff
@@ -345,6 +349,74 @@ func downloadSpecTestFixtures(csdb *build.ChecksumDB, cachedir string) string {
 	return filepath.Join(cachedir, base)
 }
 
+// hashDir iterates all files under the top-level project directory
+// computing the hash of each file (excluding files within the tests
+// subrepo)
+func hashDir() (map[string]common.Hash, error) {
+	res := make(map[string]common.Hash)
+	err := filepath.WalkDir(".", func(path string, d os.DirEntry, err error) error {
+		if strings.HasPrefix(path, filepath.FromSlash("tests/testdata")) {
+			return nil
+		}
+		if d.IsDir() {
+			return nil
+		}
+		// open the file and hash it
+		f, err := os.OpenFile(path, os.O_RDONLY, 0666)
+		if err != nil {
+			return err
+		}
+		hasher := sha256.New()
+		if _, err := io.Copy(hasher, f); err != nil {
+			return err
+		}
+		res[path] = common.Hash(hasher.Sum(nil))
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// doGenerate ensures that re-generating generated files does not cause
+// any mutations in the source file tree:  i.e. all generated files were
+// updated and committed.  Any stale generated files are updated.
+func doGenerate() {
+	tc := new(build.GoToolchain)
+	cachedir := flag.String("cachedir", "./build/cache", "directory for caching protoc/protoc-gen-go binaries.")
+
+	protocPath := downloadProtoc(*cachedir)
+	protocGenGoPath := downloadProtocGenGo(*cachedir)
+
+	preHashes, err := hashDir()
+	if err != nil {
+		log.Fatal("failed to compute map of source hashes", "err", err)
+	}
+
+	c := tc.Go("generate", "./...")
+	newPathEnv := os.Getenv("PATH") + string(os.PathListSeparator) + protocPath + string(os.PathSeparator) + "bin"
+	newPathEnv += string(os.PathListSeparator) + protocGenGoPath
+	c.Env = append(c.Env, "PATH="+newPathEnv)
+	build.MustRun(c)
+
+	postHashes, err := hashDir()
+
+	updates := []string{}
+	for path, postHash := range postHashes {
+		preHash, ok := preHashes[path]
+		if !ok {
+			updates = append(updates, path)
+		}
+		if preHash != postHash {
+			updates = append(updates, path)
+		}
+	}
+	for _, updatedFile := range updates {
+		fmt.Fprintf(os.Stderr, "updated generated file %s\n", updatedFile)
```

Comment:
Ah, we have to return an error code if anything is updated here, instead of only printing to stderr.